File: packages/core/growth.py
import math

PHI = (1 + math.sqrt(5)) / 2


def fib(n: int) -> int:
    i = int(n) if n is not None else 0
    if i < 0:
        i = 0
    a = 0
    b = 1
    for _ in range(i):
        a, b = b, a + b
    return a


def fib_floor(value: int) -> int:
    v = int(value) if value is not None else 0
    if v < 0:
        v = 0
    if v <= 1:
        return v

    a = 0
    b = 1
    while b <= v:
        a, b = b, a + b
    return a


def fib_ceil(value: int) -> int:
    v = int(value) if value is not None else 0
    if v < 0:
        v = 0
    if v <= 1:
        return v

    a = 0
    b = 1
    while b < v:
        a, b = b, a + b
    return b


def fib_nearest(value: int) -> int:
    v = int(value) if value is not None else 0
    if v < 0:
        v = 0
    lo = fib_floor(v)
    hi = fib_ceil(v)
    if (v - lo) <= (hi - v):
        return lo
    return hi


def phi_scale(value: float, steps: float = 1.0) -> float:
    if value is None or steps is None:
        return float("nan")
    return float(value) * (PHI ** float(steps))


def phi_round(value: float, steps: float = 1.0) -> int:
    return int(round(phi_scale(value, steps)))

File: apps/hive/src/growth.js
// HEADY_BRAND:BEGIN
// ╔══════════════════════════════════════════════════════════════════╗
// ║  ██╗  ██╗███████╗ █████╗ ██████╗ ██╗   ██╗                     ║
// ║  ██║  ██║██╔════╝██╔══██╗██╔══██╗╚██╗ ██╔╝                     ║
// ║  ███████║█████╗  ███████║██║  ██║ ╚████╔╝                      ║
// ║  ██╔══██║██╔══╝  ██╔══██║██║  ██║  ╚██╔╝                       ║
// ║  ██║  ██║███████╗██║  ██║██████╔╝   ██║                        ║
// ║  ╚═╝  ╚═╝╚══════╝╚═╝  ╚═╝╚═════╝    ╚═╝                        ║
// ║                                                                  ║
// ║  ∞ SACRED GEOMETRY ∞  Organic Systems · Breathing Interfaces    ║
// ║  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  ║
// ║  FILE: apps/hive/src/growth.js                                                    ║
// ║  LAYER: backend/src                                                  ║
// ╚══════════════════════════════════════════════════════════════════╝
// HEADY_BRAND:END
'use strict';

const PHI = (1 + Math.sqrt(5)) / 2;

function fib(n) {
    const i = Math.max(0, Math.floor(Number(n)));
    let a = 0;
    let b = 1;
    for (let k = 0; k < i; k++) {
        const next = a + b;
        a = b;
        b = next;
    }
    return a;
}

function fibFloor(value) {
    const v = Math.max(0, Math.floor(Number(value)));
    if (v <= 1) return v;

    let a = 0;
    let b = 1;
    while (b <= v) {
        const next = a + b;
        a = b;
        b = next;
    }
    return a;
}

function fibCeil(value) {
    const v = Math.max(0, Math.floor(Number(value)));
    if (v <= 1) return v;

    let a = 0;
    let b = 1;
    while (b < v) {
        const next = a + b;
        a = b;
        b = next;
    }
    return b;
}

function fibNearest(value) {
    const v = Math.max(0, Math.floor(Number(value)));
    const lo = fibFloor(v);
    const hi = fibCeil(v);
    if ((v - lo) <= (hi - v)) return lo;
    return hi;
}

function phiScale(value, steps = 1) {
    const v = Number(value);
    const s = Number(steps);
    if (!Number.isFinite(v) || !Number.isFinite(s)) return NaN;
    return v * Math.pow(PHI, s);
}

function phiRound(value, steps = 1) {
    const scaled = phiScale(value, steps);
    if (!Number.isFinite(scaled)) return NaN;
    return Math.round(scaled);
}

module.exports = {
    PHI,
    fib,
    fibFloor,
    fibCeil,
    fibNearest,
    phiScale,
    phiRound,
};
